File: cpp/log_reg.cpp
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/sysctl.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

double cost(double X_train[], double y_train[], double w, double b, size_t len_X){
    double cost = 0;
	double x;
	double y;
    for(size_t i = 0; i < len_X; i++){
        x = X_train[i];
        y = y_train[i];
        cost += y*log(1/(1+exp(-w*x-b))) + (1-y)*log(1 - 1/(1+exp(-w*x-b)));
    }
    return ((double)-1/(double)len_X)*cost;
}

// Return 1 means move uphill
// Return 0 means remain the same
int move_uphill(double delta_E, double T){
    if((double)rand()/(double)RAND_MAX < exp(delta_E/T))
        return 1;
    else
        return 0;
}

void simulated_annealing(double w_b[], double X_train[], double y_train[],
    size_t iterations, size_t len_X)
{
    w_b[0] = (double)rand()/(double)RAND_MAX;
    w_b[1] = (double)rand()/(double)RAND_MAX;
    double T = 1;
    double d = 0.99;
    for(size_t i = 0; i < iterations; i++){
        T = T * d;
        if(T == 0)
            return;
        double new_w = (double)rand()/(double)RAND_MAX;
        double new_b = (double)rand()/(double)RAND_MAX;
        double delta_E = cost(X_train, y_train, w_b[0], w_b[1], len_X) -
            cost(X_train, y_train, new_w, new_b, len_X);
        if(delta_E > 0){
            w_b[0] = new_w;
            w_b[1] = new_b;
        }
        else{
            if(move_uphill(delta_E, T) == 1){
                w_b[0] = new_w;
                w_b[1] = new_b;
            }
        }
    }

    return;
}

void predict(double predicted_labels[], double w_b[], double X[], size_t len_X){
    double w = w_b[0];
    double b = w_b[1];
    double x;
    size_t i;
    for(i = 0; i < len_X; i++){
        x = X[i];
        if(1/(1+exp(-w*x-b)) >= 0.5)
            predicted_labels[i] = 1;
        else
            predicted_labels[i] = 0;
    }

    return;
}

double calc_acc(double predicted_labels[], double labels[], size_t len_labels){
    int total = 0;
    for(size_t i = 0; i < len_labels; i++){
        total += (predicted_labels[i] == labels[i]) ? (1) : (0);
    }
    return (double)total/(double)len_labels;
}

int log_reg(){
	FILE *fp;
	char *line;
	size_t len = 0;
	ssize_t read = 0;

	fp = fopen("bitcoin_clean_cpp.csv", "r");
	if(fp == NULL)
		return EXIT_FAILURE;
	
	//get first line
	read = getline(&line, &len, fp);

	int num_days = -30; //offset due to no labels in last 30 rows
	while((read = getline(&line, &len, fp)) != -1){\
		num_days++;
	}
	fclose(fp);

	int train_size = floor(0.85*num_days);
	int test_size = num_days-train_size;
	double X[train_size];
	double y[train_size];
	double X_test[test_size];
	double y_test[test_size];

	fp = fopen("bitcoin_clean_cpp.csv", "r");

	read = getline(&line, &len, fp);

	char *token;
	int i = 0;
	while((read = getline(&line, &len, fp)) != -1 && i < num_days){
		if(i < train_size){
			//get timestamp
			token = strtok(line, ",");
			//get weighted price
			token = strtok(NULL, ",");
			X[i] = atof(token);
			//get labels
			token = strtok(NULL, "\n");
			y[i] = atof(token);
		}
		else{
			//get timestamp
            token = strtok(line, ",");
            //get weighted price
            token = strtok(NULL, ",");
            X_test[i-train_size] = atof(token);
            //get labels
            token = strtok(NULL, "\n");
            y_test[i-train_size] = atof(token);
		}

		i++;
	}

	fclose(fp);
    if(line)
        free(line);

	double w_b[2];
	size_t iterations = 1000;
	simulated_annealing(w_b, X, y, iterations, (size_t)train_size);

	double predicted_labels[train_size];

	predict(predicted_labels, w_b, X, (size_t)train_size);
	double train_acc = calc_acc(predicted_labels, y, (size_t)train_size);
	
	predict(predicted_labels, w_b, X_test, (size_t)test_size);
	double test_acc = calc_acc(predicted_labels, y_test, (size_t)test_size);

	printf("training accuracy = %lf\n", train_acc);
	printf("testing accuracy = %lf\n", test_acc);

	return EXIT_SUCCESS;
}

File: rust/src/log_reg.rs
extern crate rand;
use std::fs::File;
use std::io::{self, prelude::*, BufReader};
use std::str::FromStr;
use rand::distributions::{Distribution, Uniform};

const NUMDAYS:usize = 3254;
const TRAINSIZE:usize = 2765;
const TESTSIZE:usize = 489;

pub(crate) fn log_reg() -> io::Result<()> {
    let file = File::open("bitcoin_clean_rust.csv")?;
    let reader = BufReader::new(file);

	let mut x:[f64; NUMDAYS] = [0.0; NUMDAYS];
    let mut y:[f64; NUMDAYS] = [0.0; NUMDAYS];
	let mut x_test:[f64; NUMDAYS] = [0.0; NUMDAYS];
	let mut y_test:[f64; NUMDAYS] = [0.0; NUMDAYS];
	let mut i = 0;
	let mut j = 0;
	let mut flag = -1;
    for line in reader.lines() {
		if flag == -1{
			flag += 1;
			continue;
		}
		if i >= NUMDAYS{
			break;
		}
		if i < TRAINSIZE{
			for s in line?.split(","){
				//get weighted_price
				if j == 1{
					x[i] = f64::from_str(s).unwrap();
				}
				//get label
				if j == 2{
					y[i] = f64::from_str(s).unwrap();
				}
				j += 1;
			}
		}
		else{
			for s in line?.split(","){
				//get weighted_price
				if j == 1{
					x_test[i-TRAINSIZE] = f64::from_str(s).unwrap();
				}
				//get label
				if j == 2{
					y_test[i-TRAINSIZE] = f64::from_str(s).unwrap();
				}
				j += 1;
			}
		}
		j = 0;
		i += 1;
    }

	let mut w_b:[f64; 2] = [0.0; 2];
	let iterations:usize = 1000;
	simulated_annealing(&mut w_b, &mut x, &mut y, iterations, TRAINSIZE);

	let mut predicted_labels:[f64; NUMDAYS] = [0.0; NUMDAYS];

	predict(&mut predicted_labels, &mut w_b, &mut x, TRAINSIZE);
	let train_acc = calc_acc(&mut predicted_labels, &mut y, TRAINSIZE);

	predict(&mut predicted_labels, &mut w_b, &mut x_test, TESTSIZE);
	let test_acc = calc_acc(&mut predicted_labels, &mut y_test, TESTSIZE);

	println!("training accuracy = {}", train_acc);
	println!("testing accuracy = {}", test_acc);

    Ok(())
}

fn cost(x_train:&mut [f64; NUMDAYS], y_train:&mut [f64; NUMDAYS], w:f64,
	b:f64, len_x:usize) -> f64
{
	let mut cost:f64 = 0.0;
	let mut x:f64;
	let mut y:f64;
	for i in 0..len_x{
		x = x_train[i];
		y = y_train[i];
		let z = -w*x-b;
		let f = 1.0/(1.0+z.exp());
		cost += y*f.ln() + (1.0-y)*(1.0-f).ln();
	}
	return (-1.0/(len_x as f64))*cost;
}

fn move_uphill(delta_e:f64, t:f64) -> bool{
	let x:f64 = delta_e/t;
	let between = Uniform::from(0.0..1.0);
	let mut rng = rand::thread_rng();
	let p:f64 = between.sample(&mut rng);
	if p < x.exp(){
		return true;
	}
	else{
		return false;
	}
}

fn simulated_annealing(w_b:&mut [f64; 2], x_train:&mut [f64; NUMDAYS], y_train:&mut [f64; NUMDAYS],
	iterations:usize, len_x:usize)
{
	let between = Uniform::from(0.0..1.0);
    let mut rng = rand::thread_rng();
	w_b[0] = between.sample(&mut rng);
	w_b[1] = between.sample(&mut rng);
	
	let mut new_w:f64;
	let mut new_b:f64;
	let mut delta_e:f64;
	let mut t:f64 = 1.0;
	let d:f64 = 0.99;
	
	for _i in 0..iterations{
		t = t * d;
		if t == 0.0{
			return;
		}
		new_w = between.sample(&mut rng);
		new_b = between.sample(&mut rng);
		delta_e = cost(x_train, y_train, w_b[0], w_b[1], len_x) -
			cost(x_train, y_train, new_w, new_b, len_x);
		if delta_e > 0.0{
			w_b[0] = new_w;
			w_b[1] = new_b;
		}
		else{
			if move_uphill(delta_e, t){
				w_b[0] = new_w;
				w_b[1] = new_b;
			}
		}
	}
}

fn predict(predicted_labels:&mut [f64; NUMDAYS], w_b:&mut [f64; 2],
	x_train:&mut [f64; NUMDAYS], len_x:usize)
{
	let w:f64 = w_b[0];
	let b:f64 = w_b[1];
	let mut x:f64;
	let mut z:f64;
	let mut f:f64;
	for i in 0..len_x{
		x = x_train[i];
		z = -w*x-b;
		f = 1.0/(1.0+z.exp());
		if f >= 0.5{
			predicted_labels[i] = 1.0;
		}
		else{
			predicted_labels[i] = 0.0;
		}
	}
}

fn calc_acc(predicted_labels:&mut [f64; NUMDAYS], labels:&mut [f64; NUMDAYS],
	len_labels:usize) -> f64
{
	let mut total:f64 = 0.0;
	for i in 0..len_labels{
		if predicted_labels[i] == labels[i]{
			total += 1.0;
		}
	}
	return total/(len_labels as f64);
}
